action_server/src/action_server/actions/send_picture.py

```diff
@@ -0,0 +1,102 @@
+from action import Action, ConfigurationData
+import rospy
+import robot_smach_states as states
+
+from entity_description import resolve_entity_description
+
+
+class SendPicture(Action):
+    """ The Inspect class implements the action to send a picture over a whatsapp client.
+
+    Parameters to pass to the configure() method are:
+     - `target-location` (required): an entity with a segmentation area to inspect
+    """
+    def __init__(self):
+        Action.__init__(self)
+
+    class Semantics:
+        def __init__(self):
+            self.target_location = None
+
+    @staticmethod
+    def _parse_semantics(semantics_dict):
+        semantics = SendPicture.Semantics()
+        semantics.target_location = resolve_entity_description(semantics_dict['target-location'])
+
+        return semantics
+
+    class Context:
+        def __init__(self):
+            self.location_designator = None
+
+    @staticmethod
+    def _parse_context(context_dict):
+        context = SendPicture.Context()
+
+        if 'location-designator' in context_dict:
+            context.location_designator = context_dict['location-designator']
+
+        return context
+
+    def _configure(self, robot, config):
+        self._robot = robot
+
+        self.semantics = self._parse_semantics(config.semantics)
+        self.context = self._parse_context(config.context)
+
+        if self.context.location_designator is None:
+            # Request navigation action
+            self._config_result.required_context = {'action': 'navigate-to'}
+            if self.semantics.target_location is not None:
+                self._config_result.required_context['target-location'] = config.semantics['target-location']
+            return
+        # We can now assume we arrived at the place to take the picture from.
+
+        self.detect_face_state_machine = states.DetectFace(self._robot)
+
+        self._config_result.succeeded = True
+        return
+
+    def _start(self):
+        self._robot.speech.speak("I will take a picture and send it to my operator now. ")
+        r = rospy.Rate(1.0)
+        i = 0
+        result = None
+        while result is not 'succeeded' and i < 10:
+            result = self.detect_face_state_machine.execute(self._robot)
+            r.sleep()
+            i += 1
```

Comment:
This looping behavior should not be in here. It effectively breaks the cancel method, and if this kind of logic is necesarry, it should be in the state machine to begin with.